### ISSUE TITLE
Bug: Usage with generic Dictionary<string, object> and larger boolean Arrays

### ISSUE THREAD

**Issue by guidoschmidt** [rikimaru0345/Ceras#62]
Remember: The more faster I can reproduce your problem, the faster I can fix it :P

**Bug description**
We are trying to use Ceras for serialization/deserialization of custom classes that have a generic Dictionary of kind `Dictionary<string, object>` to be able to exchange a bunch of arrays of different type over UDP with Unity from [vvvv](https://vvvv.org).

I coded a quick example that shows the basic idea of what we are trying to do: [NSYNK/Ceras Issue](https://github.com/nsynkde/Ceras-Issue).

Looks like when the dict has a too large array of booleans it throws the following exception:

```
System.Exception
  HResult=0x80131500
  Message=Cannot find type System.Single[] after searching in all user provided assemblies and all loaded assemblies. Is the type in some plugin-module that was not yet loaded? Or did the assembly that contains the type change (ie the type got removed)?
  Source=Ceras
  StackTrace:
   at Ceras.SimpleTypeBinder.GetTypeFromBase(String baseTypeName)
   at Ceras.Formatters.TypeFormatter.Deserialize(Byte[] buffer, Int32& offset, Type& type)
   at Ceras.Formatters.ReferenceFormatter`1.Deserialize(Byte[] buffer, Int32& offset, T& value)
   at Ceras.Resolvers.StandardFormatterResolver.KeyValuePairFormatter`2.Deserialize(Byte[] buffer, Int32& offset, KeyValuePair`2& kvp)
   at Ceras.Formatters.CollectionFormatter`2.Deserialize(Byte[] buffer, Int32& offset, TCollection& value)
   at Ceras.Formatters.ReferenceFormatter`1.Deserialize(Byte[] buffer, Int32& offset, T& value)
   at Ceras.CerasSerializer.Deserialize[T](T& value, Byte[] buffer, Int32& offset, Int32 expectedReadLength)
   at Ceras.CerasSerializer.Deserialize[T](T& value, Byte[] buffer)
   at CerasDictionaryIssue.Program.TryWithCount(Int32 boolCount, Int32 floatCount) in E:\git\nsynk\CerasDictionaryIssues\CerasDictionaryIssue\Program.cs:line 30
   at CerasDictionaryIssue.Program.Main(String[] args) in E:\git\nsynk\CerasDictionaryIssues\CerasDictionaryIssue\Program.cs:line 41
```

Propablly we are missusing Ceras or maybe we are missing a proper configuration of it? 🤔 

**How to reproduce the bug**
See simple example: [NSYNK/Ceras Issue](https://github.com/nsynkde/Ceras-Issue)

**Platform**
- Ceras Version: 4.1.4 on net472
- We are using Unity version 2019.1.9f1, but not IL2CPP

**Comment by rikimaru0345** [comment]
Thanks for the report. I'll take a look in the next few hours. :) 

After a quick glance it seems to be some sort of bug and not a missing/wrong config.

**Comment by rikimaru0345** [comment]
Hi, just had 5 free minutes to check it out.
I can't reproduce it on the `master` branch!

Can you retry it using the latest commit?
Use the source-code directly for Unity. I've updated the guide just a few days ago (just thought I'd mention it in case there's anything unclear).

Or is there something I'm missing maybe?
I'm able to let all the 4 calls run though without the exception:
https://i.imgur.com/SLyN0Ht.png

**Comment by rikimaru0345** [comment]
I've started an appveyor/nuget build just in case one of the latest commits did in fact fix the issue.

**Comment by guidoschmidt** [comment]
Hey, thanks so much for the fast reply and suppert. Unfortunatelly it didn't fix the issue for us (in our example solution). Though it works when we clone the Ceras repo and build and execute the LiveTesting project 🤔

**Comment by guidoschmidt** [comment]
Could you try to run the solution from our example repo? Maybe we are missing a Reference or sth.

**Comment by rikimaru0345** [comment]
Will do!

**Comment by rikimaru0345** [comment]
Can't reproduce:
https://i.imgur.com/19Axs7F.png

Here's what I did:
- clone your repo
- change ceras nuger reference to the one I just built (4.1.5, pushed to nuget an hour ago)
- set breakpoint
- F5

**Comment by guidoschmidt** [comment]
Nice - thanks so much 💓 btw. we only have issues on the fourth call `TryWithCount(60, 120);`. Sorry if that wasn't clear

**Comment by rikimaru0345** [comment]
Just reporting back wanting to say that I found the issue, but...
well, something very strange is going on.
I'll just leave the log of my Immediate Window here 🤣 

```
baseTypeName == "System.Single[]"
false
string.Compare(baseTypeName, "System.Single[]")
0
baseTypeName == "System.Single[]"
false
System.Text.Encoding.UTF8.GetBytes(baseTypeName)
{byte[15]}
    [0]: 83
    [1]: 121
    [2]: 115
    [3]: 116
    [4]: 101
    [5]: 109
    [6]: 46
    [7]: 83
    [8]: 105
    [9]: 110
    [10]: 103
    [11]: 108
    [12]: 101
    [13]: 91
    [14]: 93
System.Text.Encoding.UTF8.GetBytes("System.Single[]")
{byte[15]}
    [0]: 83
    [1]: 121
    [2]: 115
    [3]: 116
    [4]: 101
    [5]: 109
    [6]: 46
    [7]: 83
    [8]: 105
    [9]: 110
    [10]: 103
    [11]: 108
    [12]: 101
    [13]: 91
    [14]: 93
```

The results of the first two command alone will probably already explain my confusion haha.

I had hoped to resolve this in a matter of minutes, but it seems like it will take a bit longer. 
I'll continue debugging and let you know as soon as I figure out more 😄

**Comment by rikimaru0345** [comment]
I barely dare to say it, but it seems you might just have helped me to find a serious bug in one of .NET Frameworks string handling methods.

I'm fully aware that the chances of finding a bug in the .NET Framework are incredibly low.
## So big disclaimer here
It might just be me being dumb or not understanding something!

I'll post my findings and an explanation here soon, and then I'll post in the microsoft corefx github (probably to the effect of them pointing out where I went wrong and then proceeding to laugh me out of there 🤣 😄).

**Comment by rikimaru0345** [comment]
Ok, here it goes:

Why is started following the strings and string-comparison in the first place
Obviously there's something wrong here:
![Immediate Window](https://i.imgur.com/76lQBMZ.png)

I then tried `string.CompareOrdinal(baseTypeName, "System.Single[]")`, which returns `18602`. But why would it?? Both strings have the same length (15) and comparing them 'char by char' also shows no difference whatsoever.

So I wondered how `CompareOrdinal` comes up with that value and explored how it works. Obviously I want to look at the code that actually gets executed, so instead of looking up the reference source code on the internet, I used dnSpy to decompile so I can be sure that's the code that's currently running. That's this exact file: 
![mscorlib](https://i.imgur.com/COJunRf.png)

I copied over the decompiled code to my project and replaced the internal parts. To access the private field `m_firstChar` I just used reflection (GetField, GetValue) and for the pointer I just wrote `fixed(char* ptr = strA)`.

After running the application again it pretty much immediately revealed how/why things went wrong here. This screenshot has all the needed information, I'll elaborate on what I think is happening below.

![Memory View](https://i.imgur.com/wktBAAp.png)

The method defines `ptr3` and `ptr4` as `char*`.
At this point they have both been incremented by 2 multiple times by the loop/check above, but by now they **already** point to the last char in their respective strings.

![StrA](https://i.imgur.com/ZLJYm4a.png)

So the last statement `return (int)(ptr3[1] - ptr4[1]);` compares stuff beyond the end of the strings.
Maybe this `[1]` is a remnant of comparing the second byte of a char, but later the code got changed from `byte*` to the current `char*` or something? Pure speculation.

Anyway, one might argue "but the decompiled code might have bugs or works differently in subtle ways", but the original `string.CompareOrdinal()` returns the identical result and `ptr3[1]` is clearly out of range with `18602 '䢪'` : 
![string compare ordinal](https://i.imgur.com/JbYJHqO.png)

**Comment by rikimaru0345** [comment]
### So what is the actual root cause for the bug you found @guidoschmidt ?
It seems (and that's just a guess) that `Assembly.GetType` also uses `string.CompareOrdinal()` (or a copy that has the exact same bug).

### So what does that mean for Ceras? 
I don't know yet.
How would I even fix this inside `Assembly.GetType` ??

Maybe strings are supposed to have `\0\0` at the end? But looking at my "literal" string in the memory view doesn't show it ending with that either.

**Comment by rikimaru0345** [comment]
I thought of a solution. I'll have time to implement it later today. Will post updates here. :)

**Comment by guidoschmidt** [comment]
Wow thanks to much for all the work and the detailed updates about that bug. It's remarkable how much effort you put in there and how fast you replied 👍 Currently we did a quick fix: encoding our boolean Arrays into byte Arrays (0 false, 128 true) which works fine so far. But I'll try keep updated here and in the bug thread about the CompareOrdinal function 🙃

**Comment by rikimaru0345** [comment]
The quick fix likely won't work reliably, but it only hast to work for a few more hours.
I'll be able to start working on my ideas to fix this at around 15:00 today (german time).

Also seems like the compare methods themselves are not the problem, but rather a bug in the 'Encoding' class of .NET (I'll update my previous post with more info on that later...)

Thanks for your patience :P :D

**Comment by rikimaru0345** [comment]
I've spent the last few hours trying out a few different approaches to fix this (without waiting for microsoft to fix the root cause), some more and some less promising.
But the least "dangerous" ones are also the slowest ones :(

- Overwriting past the end of the string: would be perfect, if we can be sure that nothing after it is actual data used by something else (which we can't)
- Copying the string into a new one (not fully reliable, not as fast)
- Injecting a fake '\0' into the data when trying to deserialize. (Very complex, need to potentially even grow the buffer because it might not have enough space left; also need to mess around with internals of the string instance to adjust the length afterwards, ...)

I'll continue trying and will probably implement a setting that can be toggled on/off for the fix (so if it causes any trouble it can be deactivated easily). 
And there's also the possibility of the microsoft guys thinking of a different way this can be hotfixed :)

**Comment by rikimaru0345** [comment]
As expected, it turned out not a bug in .NET, but instead a (very) well hidden one in Ceras.
For various reasons `Marshal.SizeOf` reports the size of a bool to be 4, while `Unsafe.SizeOf` reports 1.

The issue is fixed with f8d4db31a33397ba5c9d3f24189e59ce363e5580 and the v4.1.6 compiled from that is pushed to nuget. 😄 

Feel free to reopen when needed (or just create a new issue)! 👍

**Comment by guidoschmidt** [comment]
Super awesome 🎉 
I'll try the fix tomorrow and report back here. Thanks again for all the effort you put into this (from the whole team at NSYNK). Your quick and competent help is well appreciated.
